### PR TITLE
UI: another attempt to fix actions and signals on buttons

### PR DESCRIPTION
**Pull request by klees**
This is another attempt (besides #794) to handle the problem with actions and signals that came up with the merge of #716.

The root cause of the problem is the handling of urls as actions vs. signals in the button implementation. Both mechanism didn't know from each other, thus weren't connected in any way. I fixed that by allowing signals as actions on buttons, which seems to be correct and handling them both a little more similar.

This is not the final solution, though. The fully correct solution would imo be to introduce some signals on window that perform the page change (and maybe other window-wide actions) and use these instead of urls for the buttons. This would remove some special-casing I introduced with this PR.

I also noticed, that the current implementation of the click-signals on the buttons don't take the disabled-property of the button into account. I also noticed, that the signals and the event handlers added via `withOnLoadCode` (see e.g. show_form_in_modal.php) don't interact with each other nicely. This imo shows the need for some refinement of the signal mechanism.

### CODE REVIEW

File: src/UI/Component/Button/Button.php [hunk 1]
```diff
@@ -80,4 +82,11 @@ public function withAriaChecked();
 	 * @return 	bool
 	 */
 	public function isAriaChecked();
+
+	/**
+	 * @inheritdocs
+	 *
+	 * This will also remove a string action if there currently is one.
```
Comment (Amstutz) [inline src/UI/Component/Button/Button.php:89]: +1

File: src/UI/Implementation/Component/Triggerer.php [hunk 2]
```diff
@@ -8,6 +8,8 @@
  *
  * Provides helper methods and default implementation for components acting as triggerer
  *
+ * TODO: This is missing tests!
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Triggerer.php:11]: Can you provide those? If not, who is supposed to provide those?

File: src/UI/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# 5.2.4
+
+* ButtonFactory::standard, ButtonFactory::primary, ButtonFactory::shy, ButtonFactory::tag
+  can get a signal as second parameter instead of an action. If an url was set, the
+  previously defined action-url will be deleted when a click-signal is added. Button::getAction
+  may thus also return a list of Signals instead of an url-string.
```
Comment (Amstutz) [inline src/UI/CHANGELOG.md:6]: Kudos for keeping the changelog up to date!

**Review by Amstutz — APPROVED** [review]
Thanks a lot @klees. I approve this approach, but agree on the points you are make concerning the signaling needing some improvements. 

Note that there are some slight changes in the public interface, therefore we need to announce those at the next JF.

File: src/UI/Implementation/Component/Triggerer.php (hunk 2 above)
Comment (klees) [reply src/UI/Implementation/Component/Triggerer.php:11]: In general yes, currently I have no time for that.

The complete trait currently is missing tests, we should have requested those when that trait was initially implemented. So atm I guess no one is supposed to provide them.

**Review by chfsx — APPROVED** [review]
Although I'm not a friend of methods returning or accepting different types (strings and Signals[]) I do not have a proper solution for the moment without changing too much (such as callbacks in signals or signal-stacking). Thanks for the proposal, feel free to merge!

**Comment by matthiaskunkel** [comment]
Jour Fixe, 18 DEC 2017: We highly appreciate this PR. Please merge to trunk.

**Pull request merged**